Kolmogorov's Blossom V implementation

Toptachamann opened the pull request:
The implementation of the core features of the Kolmogorov's Blossom V algorithm.
The documentation covers features essential for the code review (will be extended further if needed). Full documentation is in the todo list.

--------------------
TODO list:
- [x] Numeric stability (have ideas how to fix, but requires careful analysis)
- [x] Correct handling of the cases when the perfect matching doesn't exist at all
- [x] Write performance tests
- [x] Full documentation about the algorithms itself
- [x] Code optimizations
- [x] Measure code coverage

simlu commented:
Hi, this looks very interesting! Do you measure coverage for your tests? What are you planning for performance testing? Ideally there would be a way to generate graphs and solve them with a different implementation and then compare performance and correctness. Looking forward to testing this out!

Toptachamann commented:
Hi, thanks!
I haven't measured the code coverage, but that's a nice idea and I included it in my task list. Firstly, I wrote test cases for Primal and Dual updaters, but those didn't cover all situations. So, I generated random instances and ran the algorithm on them. The most efficient instances were point set triangulations: they are easy to solve manually and can cause cascade sequences of blossom shrinks and expands. 
Performance testing is also scheduled for the future. Since my project is focused on a "fast algorithm" for minimum weight perfect matching, they are a task of high importance. Now the code is working but is raw a little bit. I'll fix all issues with this first version and then begin to work on speed-ups.

simlu commented:
Regarding test cases: Take a look at the test cases in the blossom code I've ported if you haven't done so. You might want to port these over (they are a nice mix of various cases). This could be in addition to your test cases in code where this adds coverage / cases or as a generic set of test cases stored in files that one can easily add to (which I think is desirable to have). 

Regarding performance: I use jprofiler for profiling and can definitely take a look at it as well when you think it's in a suitable spot for me to do that. I might also write a routine to compare with the much simpler blossom I've ported. Should give us an infinite amount of test cases (though they might not be as diverse as we want). But I'll let you refine this first. I'm also very familiar with low level optimizations in Java, so don't hesitate to ping me about language specific optimization once you identified bottlenecks. 

Will be following this with interest! Greqt work so far! L~

PS I still haven't taken a closer look at the pr code. But will that once I get a bit more time here.

Toptachamann commented:
Thanks again!
I've tested my code on your test cases from files (in_1, in_2). Do I understand the output format right? For each node **u** you print its matched node **v**. 
The problem is my code produces other matchings for both in_1 and in_2: the weights are -373390 and -1161190 respectively. The matchings in the out files have weights -7881 and -15511. Optimality checks in my code are ok and the original implementation by Vladimir Kolmogorov of Blossom V produces the same results as my code. 
In your code, you have an optimality check. So, in theory, if dual variables satisfy the dual LP formulation and complementary slackness conditions are satisfied, then a perfect matching has the minimum weight.  I wonder what's the problem.

simlu commented:
Oh, I think the weights might get normalized for performance reasons?

jsichi commented on HISTORY.md:
```diff
@@ -51,7 +51,7 @@ Changes to JGraphT in each version:
     - Optimized `PushRelabelMFImpl` implementation which drastically improves max flow computations on certain graphs. (contributed by Alexandru Văleanu) 
     - Added `RandomRegularGraphGenerator` (contributed by Emilio Cruciani)
-    - Reimplemented significantly faster version of Prim's minimum spanning tree algorithm using a FibonacciHeap (suggested by Joris Kinable, contributed by Alexandru Văleanu)
+    - Reimplemented significantly faster version of Prim's minimum spanning trees algorithm using a FibonacciHeap (suggested by Joris Kinable, contributed by Alexandru Văleanu)
```
Heh, I'm pretty sure you accidentally did a search/replace for tree->trees but set your editor scope too wide :)

Toptachamann commented:
regarding the normalization: it was mentioned in the article about Blossom V performance analysis that wider weight ranges increase the running time. In other words, integer weight range [1:10^3] is better than [1:10^6]. In my option, if the range is normalized from [1:10^6] to [10^{-3}:10^3}, it won't decrease the running time. The main thing here is the amount of dual changes made during dual update phase relative to the weight of the perfect matching. With wider weight ranges the weight of the perfect matching tend to increase (I observed that when I tested the code on random instances). But the weights can be organized in such a way that at every step only a small progress can be made. So, I don't see any benefit in weights normalization.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/Edge.java:
```diff
@@ -0,0 +1,147 @@
+ * @see KolmogorovMinimumWeightPerfectMatching
+ * @since June 2018
+ */
+class Edge {
```
Avoid using such generic names as Edge, Node, Tree, State, and Initializer since this package is not dedicated to just your algorithm.  So either use more specific names, or namespace them (e.g. in a dedicated subpackage).

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/Node.java:
```diff
@@ -0,0 +1,546 @@
+    }
+
+    /**
+     * Insert the {@code edge} into linked list of incident edges of this node in the specified direction {@code dir
```
Missing close-curly

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom_v/DualUpdater.java:
```diff
@@ -15,12 +15,12 @@
  * the Eclipse Foundation.
  */
-package org.jgrapht.alg.matching;
+package org.jgrapht.alg.matching.blossom_v;
```
Use something like `org.jgrapht.alg.matching.blossom.v5` instead.

simlu commented:
Just started running speed tests using the default options and this implementation is unfortunately significantly slower than my [port here](https://github.com/simlu/EdmondsBlossom): 4 min 35 sec vs 4 sec ---- Wondering what is causing this.

- jprofiler shows significant amount of calls to `FibonacciHeap.consolidate'
- List.get/add/set is slow and should probably be replaced with something faster (fastutil?)

Will let jprofiler run a little more and see if anything else pops up

Toptachamann commented:
Hi, 
these results are very interesting. As far as I can tell you're using some dense instances. Could you supply me with some of them? It would be interesting to test the original implementation. 

Regarding the running time issues: my implementation uses List only in grow operation. I'll fix by making the operation iterative rather than recursive. Other calls to List's operations are most probably from FibonacciHeap, which my implementation heavily uses. These calls to List operations come from the consolidate() function. It uses an ArrayList as an array, so a quick fixup would be to replace it with a regular array, also it would produce "rawtypes" and "unchecked" warning due to the generic arrays issues. A good question is whether we can speed up Fibonacci heaps in general.

d-michail commented:
@Toptachamann Could you also test with some other addressable heap? Last year I wrote a library which has a large collection of heaps in Java including for example a pairing heap. See `www.jheaps.org`. It also contains a nice addressable heap interface which allows one to swap the implementations easily.

simlu commented:
Attached the data file (this is all I have for now). You can just read it line by line and split on spaces.

[data.txt](https://github.com/jgrapht/jgrapht/files/2180968/data.txt)

Will try to do some more digging tonight.

Toptachamann commented:
Thanks for the test case. I’ve made some tests. On a sparse instance this implementation runs pretty good, on dense - no. Original implementation runs in 1.9 seconds. Interesting fact is that original implementation manages to perform only 1 shrink operation, but my implementation performs more that 2k shrinks, so this is the first point to investigate.

simlu commented:
@Toptachamann Ah, that makes perfect sense. I was a little hasty when testing it (with not enough time). However it looks like the FibonacciHeap is not very efficient. Not sure if it is worth improving though. I'll take another look when things work as expected. Which leads me to the following issue.

I'm expecting to obtain identical results for identical input. However this is not the case. Following test randomly fails with error like

```sh
java.lang.AssertionError: 
Expected :[0 97, 1 61, 10 29, 100 102, 11 90, 12 59, 13 14, 15 56, 16 17, 18 105, 19 94, 2 49, 20 103, 21 32, 22 70, 23 26, 24 72, 25 64, 27 28, 3 67, 30 31, 33 34, 35 39, 36 74, 37 50, 38 88, 4 93, 40 91, 41 80, 42 69, 43 44, 45 99, 46 84, 47 81, 48 71, 5 75, 51 52, 53 57, 54 55, 58 60, 6 7, 62 104, 63 68, 65 76, 66 85, 73 95, 77 78, 79 82, 8 9, 83 92, 86 87, 89 98, 96 101]
Actual   :[0 97, 1 61, 10 90, 100 102, 11 29, 12 79, 13 14, 15 56, 16 17, 18 105, 19 94, 2 49, 20 103, 21 32, 22 70, 23 26, 24 72, 25 64, 27 28, 3 67, 30 31, 33 34, 35 53, 36 74, 37 50, 38 88, 4 93, 40 91, 41 80, 42 69, 43 44, 45 99, 46 84, 47 81, 48 71, 5 75, 51 52, 54 55, 58 60, 59 82, 6 7, 62 104, 63 68, 65 76, 66 85, 73 95, 77 78, 8 39, 83 92, 86 87, 89 98, 9 57, 96 101]
 <Click to see difference>
```

```java
    @Test
    public void testLargeGraph() {
        Graph<Integer, DefaultWeightedEdge> graph = new DefaultUndirectedWeightedGraph<>(DefaultWeightedEdge.class);

        try {
            File f = new File("PATH/TO/FILE/data.txt");
            List<String> lines = FileUtils.readLines(f, "UTF-8");
            for (String line : lines) {
                String[] edge = line.split(" ");
                if (edge.length == 3) {
                    Graphs.addEdgeWithVertices(graph, Integer.valueOf(edge[0]), Integer.valueOf(edge[1]), Integer.valueOf(edge[2]));
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        KolmogorovMinimumWeightPerfectMatching<Integer, DefaultWeightedEdge> perfectMatching = new KolmogorovMinimumWeightPerfectMatching<>(graph);
        MatchingAlgorithm.Matching<Integer, DefaultWeightedEdge> matching = perfectMatching.getMatching();
        List<String> result = new ArrayList<>();
        for (DefaultWeightedEdge e : matching) {
            int source = graph.getEdgeSource(e);
            int target = graph.getEdgeTarget(e);
            result.add(source + " " + target);
        }
        result.sort(Comparator.naturalOrder());
        List<String> expected = new ArrayList<>();
        expected.add("0 97");
        expected.add("1 61");
        expected.add("10 90");
        expected.add("100 102");
        expected.add("11 29");
        expected.add("12 79");
        expected.add("13 14");
        expected.add("15 56");
        expected.add("16 17");
        expected.add("18 105");
        expected.add("19 94");
        expected.add("2 49");
        expected.add("20 103");
        expected.add("21 32");
        expected.add("22 70");
        expected.add("23 26");
        expected.add("24 72");
        expected.add("25 64");
        expected.add("27 28");
        expected.add("3 67");
        expected.add("30 31");
        expected.add("33 34");
        expected.add("35 53");
        expected.add("36 74");
        expected.add("37 50");
        expected.add("38 88");
        expected.add("4 93");
        expected.add("40 91");
        expected.add("41 80");
        expected.add("42 69");
        expected.add("43 44");
        expected.add("45 99");
        expected.add("46 84");
        expected.add("47 81");
        expected.add("48 71");
        expected.add("5 75");
        expected.add("51 52");
        expected.add("54 55");
        expected.add("58 60");
        expected.add("59 82");
        expected.add("6 7");
        expected.add("62 104");
        expected.add("63 68");
        expected.add("65 76");
        expected.add("66 85");
        expected.add("73 95");
        expected.add("77 78");
        expected.add("8 39");
        expected.add("83 92");
        expected.add("86 87");
        expected.add("89 98");
        expected.add("9 57");
        expected.add("96 101");
        expected.sort(Comparator.naturalOrder());
        result.sort(Comparator.naturalOrder());
        Assert.assertEquals(result, expected);
    }
```
with input from
[data.txt](https://github.com/jgrapht/jgrapht/files/2182799/data.txt)

simlu commented:
The last commit seem to have solved the inconsistencies for the test case above. I'll take a another look at this tonight and will also take a look at memory usage! Great job! It's definitely coming together!

Toptachamann commented:
Only the weights of the resulting matchings should be equal, the matchings themselves can be completely different. The weights of those two matchings are equal. If there a lot of even cycles of zero reduced cost, there will be a lot of optimal perfect matchings.

I've removed recursion from the grow operation and removed usage of external data structures. Now my code solves the first instance in 10.5 seconds on my computer, but I haven't compared that with your code. I'll do tests also with fractional matching initialization. 

By the way, your implementation uses integers and my uses doubles. That can be partially the reason for the performance difference. When compiled with double instead of integer weight, the original implementation doesn't even terminate on the first test case :)

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/PrimalUpdater.java:
```diff
@@ -284,34 +253,72 @@ public void expand(Node blossom) {
+     * </ul>
+     * <p>
+     * If the {@code manyGrows} flag is true, performs recursive growing of the tree.
+     * TODO: make this operation iterative rather then recursive
```
still valid?

Toptachamann replied:
No, need to update the javadoc

simlu commented:
My test case wasn't indicating an error in the algorithm. I do realize that there can be multiple correct solutions for a given input. However if the algorithm produces a different (random) result for identical input, it becomes very, very hard to write tests for anything that utilizes it. That's what I was pointing out with my test case.

Just tested your recent changes and there seems to be a drastic speedup. I'll have a closer look tonight!

What would it take to allow integers and doubles and initialize the algorithm with a type (using the java `T` syntax)? If we think there is a drastic speedup it might be worth it. What would it take to test speed difference double vs integer for this implementation (you could just copy my conversion to integer I think)?

simlu commented:
Didn't get to spend as much time as I would like to have today, so this will continue later. However it appears that 

"org.jgrapht.util.FibonacciHeap.consolidate"
"org.jgrapht.util.FibonacciHeap.removeMin"
"org.jgrapht.util.FibonacciHeap.delete" 

is still a bottleneck. Using plain java lists (get/add/set) certainly doesn't help performance. Improving that at least should give a decent speed-up. To be continued...

Toptachamann commented:
So, for me there is 3 major directions to work on: implement direct code optimizations, test performance with integers and test performance with different priority queue implementations.

AlexandruValeanu commented:
@Toptachamann I think that the cleanest approach would be to test your implementation using other priority queues (Fibonacci heaps). Small code optimizations should not help that much and rewriting a much more optimized FibHeap is quite challenging and takes a lot of time.

 If you find that your existing implementation works 10x faster with a different queue then you can look into how to adapt it to jgrapht (note that the existing class is used by a lot of algorithms, shortest path and spanning tree are easy examples, so a new implementation would have to be at least as efficient of all of them). 

Just my 2 cents

simlu commented:
@AlexandruValeanu Does jgrapht utilize any library to improve performance for low level data structures? Something like http://fastutil.di.unimi.it/? Might be worth a consideration as speed ups are quite drastic.

jsichi commented:
@simlu a proposal for incorporating fastutil into a new jgrapht-opt module got discussed in #459 (among many other things).  The focus there was on core data structures, but naturally it would make sense in the alg context as well.

jsichi commented:
Sure, go ahead.  Note that the plan is to keep jgrapht-core free of the fastutil dependency, so for each alg to be optimized, we'll have to decide between having the entire implementation live inside of opt (or a separate opt-alg) vs only subcomponents switchable via factories.

OK now back to our regularly scheduled Blossom V.

simlu commented:
Memory breakdown is also very interesting:
![memory](https://user-images.githubusercontent.com/1539747/42672487-1c4a9adc-861b-11e8-8c0d-ddfdb4829cde.png)

I would estimate that with a different implementation of priority queue and using fastutil for the hashmap we could at very least reduce this by a few hundret mb. To me it is also not clear why the FibonacciHeap takes up so much memory.

Do these allocations all make sense to you and can you attribute where they come from? It feels like there is some unnecessary redundancy? Looking forward to hearing your thoughts!

simlu commented:
Just a quick update: Memory wise your implementation seems to perform much worse than my port (there are some larger graphs I can no longer solve with my ram). I'll do some more digging and a direct comparison later. The challenging part is to figure out what comes from the underlying data structure and what is specific to this implementation and should / can be improved.

Toptachamann commented:
First of all, @simlu and @AlexandruValeanu, thanks for your suggestions and help.
Regarding the memory issues: the first thing to improve is to optimize 4 maps that I use to switch from generic vertices and edges to internal data structures. In the min cost flow I’m currently using the following scheme: create an array list of generic vertices and in each internal node representation store an integer position of the corresponding generic vertex. In the case we have an array of internal nodes, we can store corresponding generic vertices in the same order. Consequently, when we iterate through an array of generic vertices, we know corresponding internal nodes. The harder thing is to tell the corresponding internal node by having only a reference to the generic vertex, which I need to do a lot in the tests. So, I’ll try to come up with some idea to reduce memory usage here.
Pairing heaps use less data per node, another reason to test different heap implementation.

d-michail commented:
@Toptachamann For a quick check, you can try using the following two implementations:
 - https://github.com/d-michail/jheaps/blob/master/src/main/java/org/jheaps/tree/FibonacciHeap.java
 - https://github.com/d-michail/jheaps/blob/master/src/main/java/org/jheaps/tree/PairingHeap.java
They are pretty efficient and behind a common interface, so that they are easily swappable.

From my experience the JGraphT implementation of the Fibonacci Heap is pretty efficient! It only has a few caveats like a minor bug in deletion with minus infinity key (which I have already removed in the second version we have with the comparator) and perhaps the recursion that you have already fixed.

Toptachamann commented:
Hi, 
seems that I've managed to optimize my code. Also, I've done some benchmarking, the results are here:
[results.pdf](https://github.com/jgrapht/jgrapht/files/2202447/results.pdf). The benchmarks were performed on random complete graphs with 500 vertices. According to the results, the performance of the algorithm depends on the options. 

Also, I've generated a complete graph on 1000 vertices and tested the performance of both your and my implementations: 12.669 vs. 1.407 seconds (10 warmup iterations and 20 measurements). 

I've encountered interesting properties like the performance of my code increases by a factor of 2 after several warmup iterations. Another one is that on some graphs the performance drastically differs when the sign of the edge weights is flipped. I plan to do extensive benchmarking on various types of graphs to get more insight on how the algorithm actually works. Nevertheless, I'd like to apply some other optimizations first, cause benchmarking takes lots of time (or I can use another computer for that purpose).

simlu commented:
@Toptachamann Those results look very interesting! I didn't think that the options would impact the run-time that much! Did you do any code changes to the algorithm (nothing has been pushed to your fork yet)? Based on your tests I'll try out option7 and see how that changes run-time!

For our use case what is really hinders us is the memory usage. I've written my own implementation of Graph (using the interface and fastutil) and pass that into your code. That reduced memory usage, but not enough.

I've attached my implementation of Graph, Might make it a little easier to distinguish between where memory is used. Note that I'm storing weights as ints, but should be easy to change that.
<details>
  <summary>Click to expand</summary>

```java
package algorithm;

import it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap;
import it.unimi.dsi.fastutil.ints.Int2LongOpenHashMap;
import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
import org.jgrapht.Graph;
import org.jgrapht.GraphType;

import java.util.Collection;
import java.util.Set;
import java.util.function.Supplier;

public class LightweightUndirectedGraph implements Graph<Integer, Integer> {
    private IntOpenHashSet vertexSet = new IntOpenHashSet();
    private IntOpenHashSet edgeSet = new IntOpenHashSet();
    private Int2LongOpenHashMap edgeVertices = new Int2LongOpenHashMap();
    private Int2IntOpenHashMap edgeWeights = new Int2IntOpenHashMap();

    @Override
    public Supplier<Integer> getEdgeSupplier() {
        return () -> {
            int edge = edgeSet.size();
            edgeSet.add(edge);
            return edge;
        };
    }

    @Override
    public boolean addVertex(Integer integer) {
        return vertexSet.add((int)integer);
    }

    @Override
    public Set<Integer> edgeSet() {
        return edgeSet;
    }

    @Override
    public Set<Integer> vertexSet() {
        return vertexSet;
    }

    @Override
    public boolean addEdge(Integer sourceVertex, Integer targetVertex, Integer edge) {
        vertexSet.add((int)sourceVertex);
        vertexSet.add((int)targetVertex);
        edgeVertices.put((int)edge, (((long)sourceVertex) << 32) | (targetVertex & 0xffffffffL));
        return true;
    }

    @Override
    public Integer getEdgeSource(Integer edge) {
        return (int)(edgeVertices.get((int)edge) >> 32);
    }

    @Override
    public Integer getEdgeTarget(Integer edge) {
        return (int)edgeVertices.get((int)edge);
    }

    @Override
    public void setEdgeWeight(Integer edge, double weight) {
        edgeWeights.put((int)edge, (int)weight);
    }

    @Override
    public double getEdgeWeight(Integer edge) {
        return (double)edgeWeights.get((int)edge);
    }

    @Override
    public GraphType getType() {
        return new GraphType() {
            @Override
            public boolean isDirected() {
                return false;
            }

            @Override
            public boolean isUndirected() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isMixed() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isAllowingMultipleEdges() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isAllowingSelfLoops() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isAllowingCycles() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isWeighted() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isSimple() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isPseudograph() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isMultigraph() {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean isModifiable() {
                throw new UnsupportedOperationException();
            }

            @Override
            public GraphType asDirected() {
                throw new UnsupportedOperationException();
            }

            @Override
            public GraphType asUndirected() {
                throw new UnsupportedOperationException();
            }

            @Override
            public GraphType asMixed() {
                throw new UnsupportedOperationException();
            }

            @Override
            public GraphType asUnweighted() {
                throw new UnsupportedOperationException();
            }

            @Override
            public GraphType asWeighted() {
                throw new UnsupportedOperationException();
            }

            @Override
            public GraphType asModifiable() {
                throw new UnsupportedOperationException();
            }

            @Override
            public GraphType asUnmodifiable() {
                throw new UnsupportedOperationException();
            }
        };
    }

    @Override
    public Set<Integer> getAllEdges(Integer sourceVertex, Integer targetVertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Integer getEdge(Integer sourceVertex, Integer targetVertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Supplier<Integer> getVertexSupplier() {
        throw new UnsupportedOperationException();
    }

    @Override
    public int degreeOf(Integer vertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Set<Integer> edgesOf(Integer vertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public int inDegreeOf(Integer vertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Set<Integer> incomingEdgesOf(Integer vertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public int outDegreeOf(Integer vertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Set<Integer> outgoingEdgesOf(Integer vertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean removeAllEdges(Collection<? extends Integer> edges) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Set<Integer> removeAllEdges(Integer sourceVertex, Integer targetVertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean removeAllVertices(Collection<? extends Integer> vertices) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Integer removeEdge(Integer sourceVertex, Integer targetVertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean removeEdge(Integer edge) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean removeVertex(Integer integer) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean containsEdge(Integer sourceVertex, Integer targetVertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean containsEdge(Integer edge) {
        throw new UnsupportedOperationException();
    }

    @Override
    public boolean containsVertex(Integer integer) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Integer addEdge(Integer sourceVertex, Integer targetVertex) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Integer addVertex() {
        throw new UnsupportedOperationException();
    }
}
```

</details>

Toptachamann commented:
Pushed changes, but the won't pass maven build since I need to modify tests.

simlu commented on jgrapht-core/src/test/java/org/jgrapht/alg/matching/blossom/v5/Benchmarks.java:
```diff
@@ -0,0 +1,137 @@
+            new Options(NONE, MULTIPLE_TREE_FIXED_DELTA, true, true), //[4]
+            new Options(NONE, MULTIPLE_TREE_FIXED_DELTA, true, false), //[5]
+            new Options(NONE, MULTIPLE_TREE_FIXED_DELTA, false, true), //[6]
+            new Options(NONE, MULTIPLE_TREE_FIXED_DELTA, false, false), //[7]
```
In the test sheet (result.pdf) it says that option7 is greedy?

Toptachamann replied:
I’ve reordered them so that we have first parameter intialization type, then dual strategy, then update duals before, then after. I’ve put a second table in the results for that purpose.

simlu commented:
Interestingly enough there doesn't seem to be one technique that is best for all dense graphs. It seems that different strategies work better for different dense graphs. I certainly can't reproduce your findings with the dense graphs I use (not automatically generated) and there doesn't seem to be one strategy that works best for all of them either.

Honestly not sure how valuable it is to peruse the optimization chase down this path. Might prefer to pick the low hanging fruits first. I'll do some more tinkering around with these options later.

Toptachamann commented:
Benchmarks for triangulations of 1000 vertices: [results.pdf](https://github.com/jgrapht/jgrapht/files/2204898/results.pdf)

Toptachamann commented:
Results with fractional matching initialization: 
[results.pdf](https://github.com/jgrapht/jgrapht/files/2210807/results.pdf).  At least for triangulations, the speedup is not very high.

Toptachamann commented:
Benchmarking results with different queue implementations: 
[results_4.pdf](https://github.com/jgrapht/jgrapht/files/2223698/results_4.pdf). 
Summary: for sparse graphs, the speedup is marginal: 8-10%. For complete graphs, the speedup is pretty high: 21 - 24%. 
But we should also take into account that other code remains the same, i.e. augments, shrinks, etc. have the same running time. That's why I consider the speed up pretty good.

simlu commented:
@Toptachamann Would you say these are due to different heap types, or because of how efficiently they're implemented? 

Strictly speaking you should look at correlations for different size graphs and see which heap type scales best as the graph size increases - if you wanted to compare different heap types and not just implementations.

Toptachamann commented:
@simlu Good point, I’ll do that

d-michail commented:
Let us not reinvent the wheel here. It is well known that Pairing heaps (see https://doi.org/10.1145/214748.214759 and https://doi.org/10.1137/1.9781611973198.7) are often in practice faster than Fibonacci heaps. The important question is whether this speedup is significant enough in order to make sense to use them, or give the option to the user to use one or the other.

Regarding the actual implementations, I would say that they are all at the same level. Not over-optimized but efficient. Note that optimizing such heaps for the memory hierarchy of a modern platform, taking into account cpu caches, branch prediction, etc. is a research problem by its own.

Toptachamann commented:
Anyway, I've done this comparison, here are the results: 
[results_5.pdf](https://github.com/jgrapht/jgrapht/files/2227910/results_5.pdf). Quick summary:  Pairing heaps and Costless Meld Pairing Heap have first positions. Both scale good, the later - better.

d-michail commented:
Great, nice to here that the costless meld variant scaled better.

simlu commented:
@Toptachamann Have you tried out the [heap priority queues](http://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/PriorityQueue.html) that fastutil provides? Fastutil might make it into an [opt package](https://github.com/jgrapht/jgrapht/pull/633) soon, so providing it as an option would be reasonable and relatively easy - only if the speedup is significant that is.

A 10% - 20% speedup is very good, considering that this is only changing the heap type! Looking forward to testing this more!

Toptachamann commented:
@simlu  It isn't possible to use heaps from fastutil cause those heaps aren't mergeable. Of course, two heaps can be merged by adding all elements from the smaller one to the bigger one. Since my last benchmark showed that Costless Meld Pairing heaps scale better, the merge() operation should be pretty efficiently implemented. 

By the way, here I retested different queue implementations cause I was doubtful about the previous results, cause there was some weird slow down for pairing heaps for graphs with 200'000 edges. Now everything seems to be logical.

Another interesting fact is that the removal of maps from the implementation made the algorithm faster by 25% percent for both sparse and dense graphs.

simlu commented:
Makes sense! Thanks for the explanation!

I think your forgot to attach the file?

That's a pretty sweet speedup! We're those the changes you were planning to do wrt memory usage?

simlu commented:
**Excerpt**: ```When performing primal updates for tree T, we need to determine quickly for edge ... the corresponding edge ... (if it already exists).  One option would be to store an array of pointers at each tree. However, this would require `O(|V|2)` memory, where `|V|` is the number of trees in the beginning of the algorithm (which can be quite large).  We avoid the quadratic memory requirement by using the following scheme. [...]``` [Reference](https://pub.ist.ac.at/~vnk/papers/blossom5.pdf)

It appears that you are already not storing an index. Is that correct? I've pulled in the latest changes and am still puzzled why the memory requirement is so large. While it improved, memory usage is still an order of magnitude higher for your implementation compared to mine. Is this really just the inefficiency of the data structure?

Toptachamann commented:
This quote concerns the way we quickly determine the the TreeEdge that connects current tree and some other tree. That’s correct that I’m not storing index. In fact, it’s quite easy to avoid quadratic memory requirement in this place. I think the difference between two implementations in their memory usage is high because i’m using me objects (Nodes, Edges, Trees, Heaps) and you use mostly arrays of primitives and two IntArrayLists. There is an overhead of using objects instead of primitives since JVM allocates more memory for them. Also, since you store two dimensional array, there is also overhead because of that.
‘’’
We determined that as graph density increases, initialization time and total execution time of a graph both tend to increase. Additionally, we determined that total memory usage of the algorithm is largely dependent on the number of nodes and edges in a graph. These conclusions were not surprising to us and were anticipated. Because Blossom V is a search algorithm, it seems logical that adding more edges to traverse would lead to increases in initialization time and total execution time. Also, it seems logical that adding more nodes and edges to a graph would lead to an increase in memory usage.
‘’’ [reference](
https://userpages.umbc.edu/~gobbert/papers/REU2015Team6.pdf)
Original blossom V code has high memory usage as well, but it benefits from using C++ unions since not all variables of nodes and edges are needed at the same time.

Toptachamann commented:
Seems that I've done what I've planned to do to finish Blossom V. That's why I'm open to suggestions what else to do to finish this part of my GSoC project. 
Note: I still plan to improve documentation and rerun benchmarks but I'll do that between 6 and 14 August since that time period is dedicated for this purpose. 
Suggestions about particular places to improve Javadoc or inline comments are also welcome (in particular method lacks inline comment explanation).

simlu commented:
Sweet! I'll try to do a more complete review asap and will do some more profiling as well! But the more eyes the better! @jsichi @d-michail @AlexandruValeanu 

How is the code coverage atm? 

Would this be applicable btw? https://github.com/jgrapht/jgrapht/issues/641

Toptachamann commented:
Code coverage:
 * NodeTest - Node -> 22/24 method, 90% lines (toString() and getTrueDual() aren't covered)
 * PrimalUpdaterTest - PrimalUpdater -> 21/21 methods, 94% lines
 * DualUpdaterTest - DualUpdateer -> 6/6 method, 92% lines
 * InitializerTest - Initializer -> 7/7 methods, 99% lines
 * KolmogorovMinimumWeightPerfectMatchingTest : 
   - Node: 23/24 method, 96% lines
   - PrimalUpdater: 
   - DualUpdater: 21/21 methods, 99% lines
   - DualUpdater: 6/6 methods, 100% lines
   - Initializer: 7/7methods, 100% lines
   - KolmogorovMinimumWeightPerfectMatching: 21/35 methods, 81% lines (debug methods aren't covered). 
I used Intellij tool to test code coverage

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/Options.java:
```diff
@@ -0,0 +1,210 @@
+        sb.append(", updateDualsAfter=").append(updateDualsAfter);
+        sb.append('}');
+        return sb.toString();
+    }
```
```java
    @Override
    public String toString() {
        return "Options{initializationType=" + initializationType +
                ", dualUpdateStrategy=" + dualUpdateStrategy +
                ", updateDualsBefore=" + updateDualsBefore +
                ", updateDualsAfter=" + updateDualsAfter +
                '}';
    }
```

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/Options.java:
```diff
@@ -0,0 +1,210 @@
+         *
+         * @return the name of the dual updates strategy
+         */
+        public abstract String toString();
```
```java
        @Override
        public abstract String toString();
```

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/Options.java:
```diff
@@ -0,0 +1,210 @@
+         *
+         * @return the name of the initialization type
+         */
+        public abstract String toString();
```
```java
        @Override
        public abstract String toString();
```

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/State.java:
```diff
@@ -0,0 +1,320 @@
+
+import java.util.*;
+
+import static org.jgrapht.alg.matching.blossom.v5.Options.InitializationType.NONE;
```
Unused

simlu commented:
Discussing getting the other heap implementations into core [here](https://github.com/jgrapht/jgrapht/issues/645). Will try to get some more review time later today.

Toptachamann commented:
Thanks. I’ve pushed “watch” and I’m following all the activity

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/Node.java:
```diff
@@ -0,0 +1,531 @@
+            if (blossom.firstTreeChild == null) {
+                blossom.firstTreeChild = firstTreeChild;
+            } else {
+                Node first = firstTreeChild;
```
You are holding on to this because it could be the same as below and hence overwritten? When would this happen?

That aside, I would consider changing the name (maybe `firstTreeChildRef`?) so it doesn't conflict with Class variable `first`

Toptachamann replied:
This is rather a bug

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/Node.java:
```diff
@@ -0,0 +1,531 @@
+     */
+    Label label;
+    /**
+     * References of the first elements in the linked lists of edges that are incident to this node.
```
This description is a little confusing since the length of this is always two.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/Edge.java:
```diff
@@ -0,0 +1,173 @@
+ * @see KolmogorovMinimumWeightPerfectMatching
+ * @since June 2018
+ */
+class Edge {
```
Even though everything is now properly namespaced in the blossom.v5 package, it still might be a little confusing to have classes with very generic names such as Edge.  The reason is that in the Javadoc, there's a flat list of all top-level classes:

https://jgrapht.org/javadoc/allclasses-frame.html

(This is the source for one of the frames on the main Javadoc page.)

So someone browsing that may be misled into thinking that this is a general-purpose Edge class (until they click through to read the details).

So some short prefix for such class names is probably warranted, maybe e.g. BlossomVEdge, BlossomVTree, etc like you already have for BlossomVDebugger?

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/Node.java:
```diff
@@ -0,0 +1,531 @@
+     * @return the penultimate blossom of this node
+     */
+    public Node getPenultimateBlossom() {
+        if (blossomParent == null) {
```
I forgot to add the comment that this is a strict mode check. Ideally, when this method is invoked, blossomParent must be not null. This helps to reveal pretty subtle bugs. See, for example, Node#getOpposite(Node).

simlu replied:
I would just write
```java
if (blossomParent == null) {
  throw new RuntimeException("blossomParent is null!");
}
```

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/TreeEdge.java:
```diff
@@ -0,0 +1,226 @@
+    /**
+     * Debug variable
+     */
+    private static int ID = 0;
```
Removed this entirely since I need id for nodes and trees. This was for debugging addition of multiple tree edges between trees.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,851 @@
+            node = nodes[i].blossomParent;
+            while (node != null && !node.isMarked) {
+                if (node.dual < 0) {
+                    error += Math.abs(node.dual);
```
Now I do it in the following way: take an arbitrary node and add 1 to its dual.

Toptachamann commented:
I've used rebase to fix typos in the commit messages and to reword some of them, so you'll probably need to do ` reset --hard `.

jsichi commented:
The jheaps dependency is now in master (#652), so let the rebasing begin!

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVEdge.java:
```diff
@@ -0,0 +1,183 @@
+     * Position of this edge in the array {@code state.edges}. This helps to determine generic counterpart of
+     * this edge in constant time.
+     */
+    int pos;
```
`final int pos;`

simlu commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVInitializer.java:
```diff
@@ -0,0 +1,317 @@
+                i++;
+            }
+        }
+    }
```
Initialization is a little convoluted. I tried restructuring it, but then gave up :)

Toptachamann replied:
I've separated simple and greedy initialization into their own methods. Now it should be cleaner.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVOptions.java:
```diff
@@ -0,0 +1,213 @@
+    /**
+     * All possible options
+     */
+    public static final BlossomVOptions[] ALL_OPTIONS = new BlossomVOptions[]{
```
These can indeed be generated dynamically, but this will work only for extensions in the existing enums. This won't work after addition of a new enum option. Moreover, these options are now enumerated and this is the only existing enumeration. This is why it is somewhat easier to refer to a specific combination of options. It is also easier to determine which combination of options causes a specific test to fail. There were a lot of situations when the algorithm failed only with a particular combination.

simlu commented:
@Toptachamann Did you have a look at https://userpages.umbc.edu/~gobbert/papers/REU2015Team6.pdf? Some very interesting thoughts on parallelization!

Toptachamann commented:
Yeap, I’ve made a reference to this article above. I didn’t go deep into parallelization since I anyway won’t have time for it during GSoC project. 

By the way, there is a technique called price-and-repairs for dense graphs, which significantly speeds up the algorithm. I think, it is also prioritized task.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVNode.java:
```diff
@@ -0,0 +1,555 @@
+         * 0, then the direction of this iterator changes. If previous edge was the last incident edge, then
+         * {@code nextEdge} becomes null.
+         */
+        private void advance() {
```
We can't get rid of the currentDir variable and, therefore, avoid assigning nextDir to it. But that will slow down getDir() method. Since this iterator simply traverses all incident edges of a node, it can't be significantly optimized.

Toptachamann commented:
Hi, @simlu. I'm currently running benchmarks and finalize my GSoC project. You can read the draft of the final post [here ](https://medium.com/@timofey.chudakov/gsoc-2018-minimum-weight-perfect-matching-jgrapht-669ae59a1e). In particular, I compare our implementations and I need your name to mention you as the author of the implementation I'm benchmarking. So, could you please give this information? Also, if you know some other implementation it would be interesting to benchmark, I'm open to suggestions.

simlu commented:
Sure thing! Here is [my homepage](https://blackflux.com/) and you can find a bunch of information (including my linkedin profile).

I didn't write the java implementation from scratch. The initial version was ported as [mentioned](https://github.com/simlu/EdmondsBlossom) in the readme. The major alterations I made are around memory improvements (using integers instead of doubles) and more test cases.

I'm not aware of any other java implementations. However there is a js version and the python version I ported from (they should be very similar to my code).

The draft looks great. I'd put a tiny paragraph about what minimum weight perfect matching is and why it's useful, But that's up to you (and the people you expecting to read your blog post).

I really enjoyed following this pr btw!

Toptachamann commented:
Thanks!

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVEdge.java:
```diff
@@ -0,0 +1,183 @@
+    /**
+     * A reference to a node from the heap this edge is stored in.
+     * <p>
+     * <em>This variable doesn't need to be necessarily set to {@code null} after the edge is removed from FibonacciHeap
```
Is this comment still relevant now that you're using jheaps?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVEdge.java:
```diff
@@ -0,0 +1,183 @@
+
+    /**
+     * Returns an opposite edge with respect to the {@code endpoint}. <b>Note:</b> here we assume that
+     * {@code endpoint} is one of the current endpoints. The first enforces this rule.
```
The first what?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVEdge.java:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    /**
+     * Returns the original endpoint of this edge under the {@code endpoint}, which must be one of
```
I do not understand this sentence.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVInitializer.java:
```diff
@@ -0,0 +1,894 @@
+     * Goes through all nodes in the tree rooted at {@code root} and adds {@code eps} to the "+" nodes and
+     * subtracts {@code eps} from "-" nodes. Updates incident edges respectively.
+     *
+     * @param fibHeap the heap for storing best edges
```
Rename since using jheaps now.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVInitializer.java:
```diff
@@ -0,0 +1,894 @@
+    }
+
+    /**
+     *
```
Add description.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVNode.java:
```diff
@@ -0,0 +1,551 @@
+     */
+    boolean isOuter;
+    /**
+     * Support variable to identify the nodes which have been "processed" in some sense by the algorithm.
```
This description doesn't really explain much.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVNode.java:
```diff
@@ -0,0 +1,551 @@
+     * An edge, which is incident to this node and currently belongs to the matching
+     */
+    BlossomVEdge matched;
+    BlossomVEdge bestEdge;
```
I'm sure it's a very very good edge, but why is it the best?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVState.java:
```diff
@@ -0,0 +1,321 @@
+     * @param from the tail of the directed tree edge
+     * @param to   the head of the directed tree edge
+     */
+    public static BlossomVTreeEdge addTreeEdge(BlossomVTree from, BlossomVTree to) {
```
Wouldn't this be more appropriate as a non-static method on BlossomVTree?  (Or static if you prefer for symmetry.)  Likewise for many of the other methods below; move them to the most appropriate class instead of here.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVNode.java:
```diff
@@ -0,0 +1,551 @@
+     */
+    public BlossomVNode getPenultimateBlossom() {
+        if (blossomParent == null) {
+            return null; // strict mode, todo: remove
```
grep for todo/remove and eliminate all of them

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVTreeEdge.java:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    /**
+     * Returns the current heap of (-, +) cross-tree edges. Always returns the different from
```
What does "the different from" mean?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,889 @@
+     */
+    private DualSolution dualSolution;
+    /**
+     * BlossomVOptions used by the algorithm to getMatching the problem instance
```
"to getMatching"?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,889 @@
+    private BlossomVOptions options;
+
+    /**
+     * Constructs a new instance of the algorithm with the default options for it.
```
Should we use GraphType to validate that the input graph is indeed weighted?  Your comments mention that for unweighted graphs, it is better to use one of our other available algorithms, but perhaps there are still cases where this algorithm would be optimal?

Toptachamann replied:
We can use GraphType for this purpose, but should we throw an exception in the case the graph isn't weighted? This doesn't seem to be an error. 

This is really a good question. I'll compare Blossom V with EdmondsMaximumCardinalityMatching. I think at least memory usage will be much higher.

jsichi replied:
One way to implement a "warning" for this is to make it an exception by default, but then add an allowUnweighted flag in BlossomVOptions (default false) to allow the exception to be suppressed.  That way, an unaware developer will encounter the exception, and then decide whether they really want to go ahead with using Blossom V for this case; if so, they can pass a non-default option.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,889 @@
+     */
+    public KolmogorovMinimumWeightPerfectMatching(Graph<V, E> graph, BlossomVOptions options) {
+        Objects.requireNonNull(graph);
+        if ((graph.vertexSet().size() & 1) == 1) {
```
I think we should discriminate between easy-to-detect "invalid argument" (like this) and hard-to-detect (such as the cases detected in BlossomVDualUpdater and BlossomVInitializer, which aren't really invalid arguments.

Toptachamann replied:
Discriminate be throwing an exception of another type, or by mentioning it in the Javadoc?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,889 @@
+        if ((graph.vertexSet().size() & 1) == 1) {
+            throw new IllegalArgumentException(NO_PERFECT_MATCHING);
+        } else if (graph.getType().isDirected()) {
+            this.graph = new AsUndirectedGraph<>(graph);
```
Is this necessary?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,889 @@
+     * @return true iff the assigned dual variable satisfy the dual linear program formulation and
+     * complementary slackness conditions are also satisfied. The total error must not exceed EPS
+     */
+    public boolean testOptimality() {
```
Can we say "This method should always return true unless the algorithm implementation has a bug."?

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,889 @@
+     */
+    public double getError() {
+        lazyComputeMinimumWeightPerfectMatching();
+        BlossomVEdge edge;
```
In general, variables should be declared at the point of first use (the only exception is variables which need to be scoped outside of a loop, or if you are coding in K&R C).  This comment applies throughout this entire PR.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,889 @@
+                        node.isProcessed = true;
+                        processed.add(node);
+                    }
+                    //blossom.matched = null;
```
Eliminate all dead code.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java:
```diff
@@ -0,0 +1,889 @@
+    /**
+     * A solution to the dual linear program formulated on the  {@code graph}
+     */
+    public class DualSolution {
```
This should also be static, right?

Toptachamann commented:
I'm definitely fine with fast Javadoc fixes. 

This is a good idea to make internal classes reusable since it'll reduce code duplication and the overhead of implementing new related algorithms. I also vote for separate PRs.

jsichi commented:
OK, I pushed my edits.  I accidentally left one inline question marked with "jsichi" which you can answer via your own edit.

Here are a few nitpicky guidelines on phrasing for comments; I only made a few of the necessary edits for these:

* Be sure to use `{@code ...}` tags consistently when including code expressions in Javadoc

* Javadoc comments for methods can start with either an imperative ("Push the buzzer") or a subjectless present tense ("Pushes the buzzer"); different people use different styles.  But whatever you choose, apply it consistently.

* Javadoc comments on member variables should usually be in the form of noun phrases.  For example, "The widget which is responsible for pushing the buzzer" not "Pushes the buzzer".

* For inline comments which explain what a line or block of code does, use imperative ("Push the buzzer") or (less common) subjectlesss present ("Pushes the buzzer"), but never progressive ("Pushing the buzzer").

Toptachamann commented:
Here are the performance results:

With weighted graphs:

| Test type            | Blossom V, ms/op | Edmonds Blossom, ms/op |
|----------------------|------------------|------------------------|
| triang. 1000 points  |         1.713    |        1.672           |
| triang 3000 points   |         5.683    |        6.662           |
| triang 5000 points   |         10.947   |        13.482          |
| triang 5000 points   |         30.566   |        38.080          |
| complete 100 nodes   |         1.012    |        0.045           |
| complete 300 nodes   |         21.165   |        0.515           |
| complete 500 nodes   |         96.929   |        2.255           |

With unweighted graphs:

| Test type            | Blossom V, ms/op | Edmonds Blossom, ms/op |
|----------------------|------------------|------------------------|
| triang. 1000 points  |         0.601    |        1.757           |
| triang 3000 points   |         2.335    |        6.648           |
| triang 5000 points   |         4.742   |        13.208          |
| triang 5000 points   |         12.510   |        38.179          |
| complete 100 nodes   |         1.126    |        0.046           |
| complete 300 nodes   |         21.033   |        0.519           |
| complete 500 nodes   |         99.985   |        2.308           |

A quick summary: Blossom V is faster on sparse graphs and slower on dense graphs.

simlu commented:
Which implementation did you use? The one from jgrapht? I didn't know it supported weighted graphs?

Toptachamann commented:
Sorry for misterious results, it doesn’t. I included them since I run all benchmarks with both algorithms

simlu commented:
@jsichi Should be documented somewhere in a readme though. When to use which alg

jsichi commented:
@simlu I agree

jsichi commented:
OK, I made a few followup edits.  I'll merge this one in a few days unless there are more comments.

Toptachamann commented:
Here is the performance comparison with LEMON implementation:

| Test type         | Blossom V, ms/op | LEMON implementation, ms/op |
|-------------------|------------------|-----------------------------|
|  Triang. 1000     |     1.5          |       3.1                   |
|  Triang. 5000     |     10.2         |       16.4                  |
|  Complete 300     |     18.6         |       37                    |
|  Complete 300     |     86.7         |       149                   |

According to these results, this implementation outperforms implementation from LEMON. This is quite surprising since Java usually works slower than C++.

jsichi commented:
Awesome, merging now!
